procedure/pull-request-merge.rst

```diff
@@ -17,133 +16,430 @@ Memory Pool System pull request merge procedure
+
+   Alternatively, you could undo your merging work::
+
+     git reset --hard perforce/master
+
+   then go back to merging (step 4).
+
+8. Optionally, if and *only if* the Perforce push (step 7) succeeded,
+   you can also push to GitHub::
+
+     git push github master branch/2023-01-06/speed-hax
+
+   If you don't do this, then within `30 minutes
+   <https://info.ravenbrook.com/infosys/robots/gitpushbot/etc/crontab>`_
+   check that the merge appears in `the commits in the Ravenbrook MPS
+   repo on GitHub <https://github.com/Ravenbrook/mps/commits/master>`_.
+
+   If they do not appear:
+
+   1. Check email for error messages from gitpushbot and resolve them.
+
+   2. Check (or ask a sysadmin to check) that gitpushbot is running
+      on Berunda and restart it if necessary, or ask a sysadmin to do
+      this.
+
+.. _Fetch the pull request branch: https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/reviewing-changes-in-pull-requests/checking-out-pull-requests-locally#modifying-an-inactive-pull-request-locally
+
+
+5. Rationale
+------------
+
+This section explains why the procedure is like it is.  It's intended
+for people who want to vary the procedure on the fly, or make
+permanent changes to it.  In the latter case, update this section!
+
+
+5.1. Why not rebase or squash merge?
+------------------------------------
+
+We would like to avoid rewriting history and the destruction of
+information on the grounds that it destroys information that could be
+important to the engineering of the MPS, such as tracking down
+defects, comprehending the intention of changes.  So want to
+discourage rebasing or squashing.
+
+We want to avoid fast-forwards of master.  A fast-forward means there
+is no commit that records the fact that there has been a merge, by
+whom, from where, for what purpose, etc.  It discards that
+information.  Therefore we want to discourage fast-forwards of master
+in favour of merges.  (Annoyingly, GitHub only provides `branch
+protection that enforces the opposite
+<https://docs.github.com/en/repositories/configuring-branches-and-merges-in-your-repository/defining-the-mergeability-of-pull-requests/about-protected-branches#require-linear-history>`_!)
+See also `5.3. Why the "durable" branch names?`_.
+
+We also want to avoid `squash merges
+<https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/incorporating-changes-from-a-pull-request/about-pull-request-merges#squash-and-merge-your-commits>`_.
+A squash merge compresses development history into a single commit,
+destroying the record of what happened during development and the
+connection to the branch.
+
+The main motivation for fast-forwards and squashes appears to be to
+simplify the branching history so that it's easier to understand.
+Better tools and interfaces are no doubt required for analysing Git
+history.  These will emerge.  And they will be able to analyse the
+history that we are creating today.
```

Comment:
You've written "appears to be" here which suggests that you are skeptical about whether this is really people's motivation. Assuming that's not what you intended to convey, I suggest re-wording the sentence, for example

> Many developers use fast-forwards and squashes to simplify the branching history so that it's easier to understand.

Then you can argue the case for "preserving the history is more valuable than being easily able to understand it" without the distraction of appearing to doubt that developers are telling the truth when they say that simplifying the history makes it easier to understand.

As an example of this trade-off, consider commit a10d3d0a08b34ccf2c86fc6e0ab535b78adaff8e. I find this commit quite hard to understand: the commit message says "Merge branch/2023-01-10/travis-freebsd to get Travis to test this branch on FreeBSD" but the commit also contains changes relating to Windows and to the pull request procedure. Is this commit providing enough valuable information about the development history to justify the confusion it is likely to cause for future developers coming across it in `git blame` or wherever?

The claim that "better tools will emerge" reminds me of the joke about the "sufficiently smart compiler" that would absolve us of the need to write efficient code. Or, taking the claim seriously, it seems to privilege those future developers with the better tools over the current developers with the tools we have now.

Reply:
> You've written "appears to be" here which suggests that you are skeptical about whether this is really people's motivation.

I'm not skeptical, but I'm struggling to get a grip on common Git practices, the consequences they might have for our migration, our future work, and how they might affect the MPS quality in the long term.  It's a lot to think about.  Nonetheless, my expression of that doesn't belong in this procedure.

> "sufficiently smart compiler" that would absolve us of the need to write efficient code

That *does* happen though.  We need a lot fewer macros now that inlining is smarter, for example.  The MPS could do with losing a few.  What hasn't changed is the need to keep an eye on complexity.

> Is this commit providing enough valuable information about the development history to justify the confusion it is likely to cause for future developers coming across it in `git blame` or wherever?

I do not think we are in a position to judge what future people might find useful.

I know I'm in a position to judge what I'm finding useful now in untangling what's happened to the MPS: complete history of everything ever done (and why, if possible).  Unfortunately, some of it has been deleted by rebasing, and now I have to either ask people what they did and why, or guess.  We have unresolved divergence between Perforce, GitHub, and our clients' Git caused by this stuff.

Reply:
> I suggest re-wording the sentence

Done in 52c1d3a4b .  Thank you.

Reply:
Resolving this, though the more general debate can continue elsewhere!